lib/transactions-common.js

```diff
@@ -1426,8 +1426,10 @@ Transact.prototype._processTransaction = function (txid, description, items, con
   
   // First, need to iterate over the changes that are going to be made and make sure that,
   // if there are hard removes, the db version of the doc gets stored on the transaction
+  // We only do this step if the remove is not instant - otherwise the doc would already have
+  // been removed from the collection by this point.   
   _.each(items, function (item, index) {
-    if (item.action === "remove" && item.hardDelete) {
+    if (item.action === "remove" && item.hardDelete && !item.instant) {
       // Get the existing doc and store it in the transaction record
       // We overwrite the temporary version of the doc from an instant remove on the client
       // Because chances are that the whole document was not available on the client
```

Comment:
Having an incomplete document for instant removes from the client is not only an unsolved problem, it's a near-intractable problem, as the remove is not being handled by the package, but by native methods (using allow/deny).  If we're happy to wait for a server round trip via a method to fetch the whole document to place in the transaction record, that would work - but optimistic updates would be gone and there would be a big, long, lock-up on the client if there are multiple removes.

I think it's acknowledged in the docs that instant client side removes could result in data corruption/loss due to unpublished fields.